### PR TITLE
OMWorld: Remove BasicLock* restrictions

### PR DESCRIPTION
**Pull request by xmas92**
This is based on #159 

The main idea is to remove the restrictions where the `BasicLock*` may be missing when calling into the runtime. This is only an issue on x86_32. For now simply call into the runtime at the places where the registers are lacking. `quick_enter` should take care of the cases where we otherwise would have skipped calling into the runtime, as x86_32 calls into the runtime regardless (to get the thread) this may even be performance neutral.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Reviewers
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to [5c2981ad](https://git.openjdk.org/lilliput/pull/160/files/5c2981adc5f8bdf87e3c4fd6810f68cc6798f094)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/lilliput.git pull/160/head:pull/160` \
`$ git checkout pull/160`

Update a local copy of the PR: \
`$ git checkout pull/160` \
`$ git pull https://git.openjdk.org/lilliput.git pull/160/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 160`

View PR using the GUI difftool: \
`$ git pr show -t 160`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/lilliput/pull/160.diff">https://git.openjdk.org/lilliput/pull/160.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/lilliput/pull/160#issuecomment-2072116014)

### CODE REVIEW

**Review by rkennke — APPROVED** [review]
Nice! Looks reasonable. Yeah I've been thinking the same: if we need to call runtime to fetch the Thread* anyway, we might as well do all the locking stuff there.

File: src/hotspot/cpu/x86/c1_MacroAssembler_x86.cpp [hunk 1]
```diff
@@ -66,10 +67,12 @@ int C1_MacroAssembler::lock_object(Register hdr, Register obj, Register disp_hdr
   if (LockingMode == LM_LIGHTWEIGHT) {
 #ifdef _LP64
     const Register thread = r15_thread;
+    lightweight_lock(disp_hdr, obj, hdr, thread, tmp, slow_case);
 #else
-    const Register thread = disp_hdr;
+    movptr(hdr, Address(obj, oopDesc::mark_offset_in_bytes()));
```
Comment (stefank) [inline src/hotspot/cpu/x86/c1_MacroAssembler_x86.cpp:72]: Maybe add a comment about this null-check?

File: src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp [hunk 2]
```diff
@@ -1208,41 +1208,26 @@ void C2_MacroAssembler::fast_unlock_lightweight(Register obj, Register reg_rax,
...
-        orptr(reg_rax, Address(monitor, ObjectMonitor::cxq_offset()));
-        jcc(Assembler::notZero, check_successor);
+      Label recursive;
 
-        // Release lock.
-        movptr(Address(monitor, ObjectMonitor::owner_offset()), NULL_WORD);
-#else // _LP64
-        Label recursive;
+      // Check if recursive.
+      cmpptr(Address(monitor,ObjectMonitor::recursions_offset()),0);
```
Comment (stefank) [inline src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:1215]:
```suggestion
      cmpptr(Address(monitor,ObjectMonitor::recursions_offset()), 0);
```

**Review by coleenp — COMMENTED** [review]
This looks applicable to mainline, can it be pushed there?

**Comment by xmas92** [comment]
> This looks applicable to mainline, can it be pushed there?

Yes. 

There are some other changes that could be split out into upstream patches as well. The plan for now is to get this patch queue in, rebase and squash together into one OMWorld commit and then look at what can / should be broken out into upstream patches.

The hope is that after this patch queue the runtime changes are close to final (with exception of the hash table grow and shrinking code).

**Pull request closed without merging**